src/ucp/core/ucp_request.h

```diff
@@ -113,6 +113,7 @@ struct ucp_request {
             size_t                length;   /* Total length, in bytes */
             uct_memory_type_t     mem_type; /* Memory type */
             ucp_send_callback_t   cb;       /* Completion callback */
+            void                  *application_cookie; /* Value to carry context to the cb */
```

Comment:
@yosefe I think this breaks ABI backward compatibility, unless we have some protection (and I don't see one)

Reply:
@shamisp The internal structure of the ucp_request isn't exposed outside of the ucx libraries (or across the network), so I don't think this will break ABI backward compatibility.

Reply:
@tjcw We expose it through `ucp_context_query` in `ucp_context_attr_t`. But this is a runtime query and people should not base their request allocation based on static values. So I think we okay here.

Reply:
If this is active message specific, any particular reason why it is not part of am:
```
               struct {
                    uint16_t am_id;
                    uint64_t message_id;  /* used to identify matching parts
                                             of a large message */
                    unsigned flags;
                } am;
```

Reply:
@tjcw as we discussed, it's possible to add user data to request after it's allocated by UCP by reserving extra space for it in ucp_init() (params::request_size)

Reply:
I agree with @yosefe. @tjcw I can point you to an example where I did something similar to what I think you want.

Reply:
@snyjm-18 I just stumbled across your comment and have a similar issue to the one initially faced in this PR. It would be great if you could share the pointer publicly for future reference.